pkg/reconciler/integration/defaults_test.go

```diff
@@ -0,0 +1,151 @@
+// limitations under the License.
+
+package integration
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/google/kf/pkg/kf/testutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/yaml"
+)
+
+// DefaultTest defines validation for defaulting webhooks. The Resource will
+// be created and all paths will be tested for equality.
+type DefaultTest struct {
+	Path              string                 `json:"-"`
+	TestEqualityPaths []string               `json:"testEquality"`
+	Resource          map[string]interface{} `json:"resource"`
+}
+
+func (dt *DefaultTest) GetUnstructured() *unstructured.Unstructured {
+	return &unstructured.Unstructured{
+		Object: dt.Resource,
+	}
+}
+
+func (dt *DefaultTest) GetGvr() schema.GroupVersionResource {
+	obj := dt.GetUnstructured()
+	gvk := obj.GroupVersionKind()
+
+	return schema.GroupVersionResource{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+
+		// We can autosidscover this in the future from the API Server like
+		// Kubernetes does, but in the meantime all resources we deal with follow
+		// this formula.
+		Resource: strings.ToLower(gvk.Kind) + "s",
+	}
+}
+
+func loadTests(t *testing.T) []*DefaultTest {
+	t.Helper()
+
+	tests, err := filepath.Glob(filepath.Join("testdata", "defaults", "*.yaml"))
+	testutil.AssertNil(t, "defaults err", err)
+
+	// Sort so the tests are deterministic as different filesystems may traverse
+	// their directories in different ways.
+	sort.Strings(tests)
+
+	var out []*DefaultTest
+
+	for _, testPath := range tests {
+		testBytes, err := ioutil.ReadFile(testPath)
+		testutil.AssertNil(t, "read error", err)
+
+		defaultTest := &DefaultTest{}
+		err = yaml.Unmarshal(testBytes, defaultTest)
+		testutil.AssertNil(t, "unmarshal error", err)
+		defaultTest.Path = testPath
+
+		out = append(out, defaultTest)
+	}
+
+	return out
+}
+
+// TestLoadTests validates that tests can be loaded from disk
+// this test will run even if integration tests are not enabled.
+func TestLoadTests(t *testing.T) {
+	loadTests(t)
+}
+
+func TestIntegration_ApplyDefaults(t *testing.T) {
+	for _, tc := range loadTests(t) {
+		t.Run(tc.Path, func(t *testing.T) {
+			testutil.RunKubeAPITest(t, func(ctx context.Context, t *testing.T) {
+				testutil.WithNamespace(ctx, t, func(namespace string) {
+					testutil.WithDynamicClient(ctx, t, func(client dynamic.Interface) {
+						// pass
+						dyn := client.Resource(tc.GetGvr()).Namespace(namespace)
+
+						obj := tc.GetUnstructured()
+						obj.SetNamespace(namespace)
+						expected := obj.DeepCopy()
+
+						actual, err := dyn.Create(obj, metav1.CreateOptions{})
+						testutil.AssertNil(t, "creation error", err)
+
+						for _, path := range tc.TestEqualityPaths {
```

Comment:
Do we expect the object to change at all? Why can't we simply use `reflect.DeepEqual`?

Reply:
I added a comment inline about this too. We do expect some things to change like status, UID, and creation timestamp, while others might be non-deterministic like early versions of Knative Serving set a generation field in the spec.

Reply:
Cool, that makes sense!